Several VR fixes, DX10 Support

retroluxfilm opened the pull request:
I could restore the VR functionality(DX11) in the latest source code to work again after the render target rework (commit ebf94113 ) It was caused that the backbuffer_resolved was not set on the vr submit for DX11.

After it tool a long time today to find the culprit I could start to work on serveral issues I encountered on VR with reshade:
- Fixed round errors would cause that the VR backbuffer is recreated very often and causing the left eye to flicker or black out completly
- Fixed OpenGL implementation pass the proper texture type. Now Google Earth works with reshade and other opengl VR implementations
- Fixed texture format typless check on DX11 and DX12 causing to permanently recreate the VR backbuffer on some games.

There are still things left I want to fix but they might need more time as i still need to get a better understanding of the code.
- Game crash when the "preprocessor definitions" are openend through the Steam VR dashboard reshade UI
- Reshade effect on off flicker caused by dynamic resolution switching like HL-Alyx does. This is certainly an issue of a frame delay when the runtime gets reset and initialized again after each resolution switch. Not sure on how to fix that properly.

Let me know if the code is ok like that or if you need some adjustments made to it.

retroluxfilm commented:
That one is fixed now as well:
- Game crash when the "preprocessor definitions" are openend through the Steam VR dashboard reshade UI

retroluxfilm commented:
More additions:
- Implemented DX10 open VR support that allows Raceroom to run with reshade

retroluxfilm commented:
Thanks for taking the time to thoroughly review the code and comment on how it can be improved. This really helps to get future pull request to get better.  I am still pretty rusty in C++. ;)

During the next week I will find time to go through them and improve/test it so you have an easier time to merge with hopefully less clean up etc. to do.

retroluxfilm commented on source/openvr/openvr.cpp:
```diff
@@ -29,13 +32,112 @@ static inline vr::VRTextureBounds_t calc_side_by_side_bounds(vr::EVREye eye, con
+/*
+Get the device api based on the texture type
+*/
+static inline boolean isD3D10(void *pTexture)
```
I actually had a simpler approach before similar to yours but even DX11 textures could be cast to dx DX10 without failure. I could only go through the reshade device impl stored in the private data to really find if it is the right render api. Because of the complexity of the the check it will use the swapbuffer when it is resolved to speed things up.

crosire replied:
Mmh. I think these days the D3D10 driver is really just a wrapper around D3D11, so it may be that querying a D3D10 texture can also give you the D3D11 one.
In that case this should still be possible to do, just need to first query for `ID3D10Texture2D` and then for `ID3D11Texture2D` (I don't think the D3D11 driver gives you the D3D10 object), rather than in the order shown above (did it in this order, since there are vastly more D3D11 games out there, so would have been nice to avoid the cost of querying for D3D10 for those).

retroluxfilm replied:
Sadly it does. I tried it first to cast to the 'ID3D10Texture2D' and even when it was a DX11 one it passed through. The previous code crashed on DX10 accessing the proxy fetched from the device. All other stuff went before went through in the dx11 submit call.

crosire replied:
I see. I'm a bit confused, since the OpenVR headers state that `TextureType_DirectX` should be `ID3D11Texture2D` (https://github.com/ValveSoftware/openvr/blob/master/headers/openvr.h#L158). So not sure how that can be `ID3D10Texture2D`. But maybe that is just documented wrong.

It does look like the current code should crash in D3D11 though. When a `ID3D11Texture2D` object is passed into `isD3D10` it's statically cast to `ID3D10Texture2D`, even though it is not that. This by sheer luck happens to work because both `ID3D10Texture2D` and `ID3D11Texture2D` have the `GetDevice` method at the same vtable index (3) and since the vtable pointer is stored at the beginning it will make use of the `ID3D11Texture2D` method regardless of the cast (the compiler translates ```dxTexture->GetDevice(&device);``` to ```void **vtable = *(void***)dxTexture; ((void(*)(void *, void **))vtable[3])(dxTexture, &device);```). But now `device` actually points to a `ID3D11Device` object and not a `ID3D10Device` object as the code assumes. This means the `GetPrivateData` call should crash, since it is located at a different vtable index in `ID3D10Device` (76) vs `ID3D11Device` (34). There is no entry for index 76 in the `ID3D11Device` vtable, so that will call an invalid address. Only way to ensure you get the correct object would be to first `QueryInterface` on the device (since `QueryInterface` is always at vtable index 0 and therefore safe to call).

But I think this can be abused correctly like this (assuming `QueryInterface` on a D3D11 device does not return a D3D10 device, and from a quick test of mine this does not appear to be the case):
```c++
static vr::EVRCompositorError on_submit_d3d10(vr::EVREye eye, ID3D10Texture2D *texture, const vr::VRTextureBounds_t *bounds, vr::EVRSubmitFlags flags,
	std::function<vr::EVRCompositorError(vr::EVREye eye, void *texture, const vr::VRTextureBounds_t *bounds, vr::EVRSubmitFlags flags)> submit)
{
	com_ptr<ID3D10Device> device;
	texture->GetDevice(&device);

	D3D10Device *device_proxy = nullptr; // Was set via 'SetPrivateData', so do not use a 'com_ptr' here, since 'GetPrivateData' will not add a reference
	if (UINT data_size = sizeof(device_proxy);
		FAILED(device->GetPrivateData(__uuidof(D3D10Device), &data_size, reinterpret_cast<void *>(&device_proxy))))
		goto normal_submit; // No proxy device found, so just submit normally
	...
}
static vr::EVRCompositorError on_submit_d3d11(vr::EVREye eye, ID3D11Texture2D *texture, const vr::VRTextureBounds_t *bounds, vr::EVRSubmitFlags flags,
	std::function<vr::EVRCompositorError(vr::EVREye eye, void *texture, const vr::VRTextureBounds_t *bounds, vr::EVRSubmitFlags flags)> submit)
{
	com_ptr<ID3D11Device> device;
	texture->GetDevice(&device); // 'GetDevice' is at the same vtable index for 'ID3D10Texture2D' and 'ID3D11Texture2D', so this happens to work for either case

	if (com_ptr<ID3D10Device> device10;
		device->QueryInterface(&device10) == S_OK)
		// Whoops, this is actually a D3D10 texture, redirect ...
		return on_submit_d3d10(eye, reinterpret_cast<ID3D10Texture2D *>(texture), bounds, flags);

	D3D11Device *device_proxy = nullptr; // Was set via 'SetPrivateData', so do not use a 'com_ptr' here, since 'GetPrivateData' will not add a reference
	if (UINT data_size = sizeof(device_proxy);
		FAILED(device->GetPrivateData(__uuidof(D3D11Device), &data_size, reinterpret_cast<void *>(&device_proxy))))
		goto normal_submit; // No proxy device found, so just submit normally
	...
}

...

case vr::TextureType_DirectX:
	return on_submit_d3d11(eEye, static_cast<ID3D11Texture2D *>(pTexture->handle), pBounds, nSubmitFlags, submit_lambda);
```

retroluxfilm replied:
No you are right that openVR has dokumented that it is expected to get a DX11 texture. But they also have DX10 support and even worked on DX9 but that implementation never got finished. Raceroom might be the only game using DX10 for VR. Was surprised that it was possible/supported as well.

If your example code works it would be more slick and cleaner solution. But I think i tried to use QueryInterface() and it did not work as expected. But I might not did it correctly. I will try it again and when it works rework the code.

crosire replied:
Would be interesting to know if it's actually a `ID3D10Texture2D` object passed in. You can check this in the Visual Studio debugger by putting a breakpoint in the submit function and when that breaks, entering `((IUnknown*)texture)->__vfptr` in the Watch window. The Value column will then show you the name of the vtable (when symbols are loaded).

If this is actually a `ID3D11Texture2D`, then this is
```
{d3d11.dll!const CLayeredObjectWithCLS<class CTexture2D>::CContainedObject::`vftable'{for `CDeviceChild<struct ID3D11Texture2D>'}} {...}
```
If this is actually a `ID3D10Texture2D`, then this is something else, e.g.
```
{d3d11.dll!const CLayeredObjectWithCLS<class CTexture2D>::CContainedObject::`vftable'} {...}
```

retroluxfilm replied:
Damn Raceroom seems very twitchy and crash quickly when attaching the debugger... On beatsaber debugging was no issue at all. Before it crashes it also tells that the brakepoints does not match the code. Are there some tricks for this or a way to output the ((IUnknown*)texture)->__vfptr into the log file?

retroluxfilm replied:
Great it works by fetching the device interface as you suggested. Simplified this now. Thx.

retroluxfilm commented on source/d3d10/d3d10_impl_swapchain.cpp:
```diff
@@ -110,6 +113,20 @@ bool reshade::d3d10::swapchain_impl::on_init()
+
+		for (unsigned int i = 0; i < add_references; ++i)
+			_backbuffer->AddRef();
+	}																																	 
```
It might not be needed. I just thought that the part was missing on releasing the buffers properly. I was unsure if the code was only related to the comment near to it.

retroluxfilm replied:
Removed it for DX10

retroluxfilm commented on source/d3d11/d3d11_impl_swapchain.cpp:
```diff
@@ -213,15 +213,22 @@ bool reshade::d3d11::swapchain_impl::on_layer_submit(UINT eye, ID3D11Texture2D *
 
-	if (target_width != _width || region_height != _height || convert_format(source_desc.Format) != _backbuffer_format)
+	//convert the source format to the typeless format
+	const api::format convertedSourceFormat = api::format_to_typeless(convert_format(source_desc.Format));
```
Original version is now stored instead of the typeless

retroluxfilm commented:
Ok I adjusted all suggestions and added another small usability fix that the return key is passed to the VR UI to confirm values. Tested a few games out again to ensure everything works as before.

From my end you can review and merge, clean up it you like.
